src/Lean/Elab/Tactic/Change.lean

```diff
@@ -0,0 +1,47 @@
+/-
+Copyright (c) 2023 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import Lean.Meta.Tactic.Replace
+import Lean.Elab.Tactic.Location
+
+/-!
+# Implementation of the `change` tactic
+-/
+
+open Lean Elab Tactic Meta
+
+/-- `change` can be used to replace the main goal or its local
+variables with definitionally equal ones.
+
+For example, if `n : Nat` and the current goal is `⊢ n + 2 = 2`, then
+```lean
+change _ + 1 = _
+```
+changes the goal to `⊢ n + 1 + 1 = 2`. The tactic also applies to the local context.
+If `h : n + 2 = 2` and `h' : n + 3 = 4` are in the local context, then
+```lean
+change _ + 1 = _ at h h'
+```
+changes their types to be `h : n + 1 + 1 = 2` and `h' : n + 2 + 1 = 4`.
+
+Change is like `refine` in that every placeholder needs to be solved for by unification,
+but you can use named placeholders and `?_` where you want `change` to create new goals.
+
+The tactic `show e` is interchangeable with `change e`, where the pattern `e` is applied to
+the main goal. -/
```

Comment:
This seems to needlessly increase our API surface. Why can't we have just one of `change` and/or `show`?

I don't think it's a good user experience to ask why a given operator seems to just be a special case of another, and then find out "oh, yeah, this is some historical thing". We'll eventually be there for reasons of compatibility, but let's keep things consistent as long as we have users who will put up with it!

Reply:
```suggestion
/-- `change` can be used to replace the main goal or its hypotheses with
different, yet definitionally equal, goal or hypotheses.

For example, if `n : Nat` and the current goal is `⊢ n + 2 = 2`, then
```lean
change _ + 1 = _
```
changes the goal to `⊢ n + 1 + 1 = 2`.

The tactic also applies to hypotheses. If `h : n + 2 = 2` and `h' : n + 3 = 4`
are hypotheses, then
```lean
change _ + 1 = _ at h h'
```
changes their types to be `h : n + 1 + 1 = 2` and `h' : n + 2 + 1 = 4`.

Change is like `refine` in that every placeholder needs to be solved for by unification,
but using named placeholders or `?_` results in `change` to creating new goals.

The tactic `show e` is interchangeable with `change e`, where the pattern `e` is applied to
the main goal. -/
```

Where are named placeholders described? A quick grep of the Lean source doesn't turn much up. Can we add an example here?

Reply:
(that last mega-suggestion is mostly just written style harmonization/consistency with other docstrings)

Reply:
`show T` and `change T` are the same, but they each generalize in different directions.

* `show T` is also a term constructor `show T from e` / `show T by tac`, and we like to have tactics which mimic the syntax of their term variants. (I'm not a big fan of the use of `from` in this syntax and would prefer `:=`, but that's another matter.)
* `change T` also supports `change T at h` and (not upstreamed) `change e1 with e2 (at h)?`, which allow defeq rewriting in the goal or a hypothesis. While one could use `show` here instead, the resulting tactic `show e1 with e2` doesn't make a whole lot of sense grammatically (given that it means to replace occurrences of `e1` in the goal with `e2`).

Reply:
An alternative tactic design for `change e1 with e2 at h` is `drw [e1 = e2] at h`. This is using the same syntax as `rw` with a `d` prefix for "definitional" similar to `dsimp` and others, but it takes equalities directly instead of proofs (the equivalent of `show e1 = e2 from rfl` in a `rw` proof). I don't see any obvious analogue for the `change T at h` though, except `drw [_ = T] at h` which has some unnecessary syntax.

Reply:
OK, this makes sense - one is not more general than the other.

I think we should update the docstrings of both to make this more clear, then, using text similar to @digama0's comment.

I can have a go, but I want to check my understanding first: there's no context in a tactic proof where `show` would be applicable but `change` would not be, right? In other words, we can consider `show` to be something you can use if you want to, but `change` is the appropriate default for people who want to accomplish this task?

Reply:
Yes, you can always use `change` in place of `show`. But as a matter of practice `show` is considerably more popular than `change` for the case where it is applicable. (I think; it's a bit tricky to directly get evidence using grep, because `show` has lots of other uses.)